### ISSUE TITLE
Can't sign up after changing the authentication from local to email

### ISSUE THREAD

**Issue by zacker330** [wasp-lang/open-saas#13]
**Describe the bug**

The email auth is not working.

**To Reproduce**
1. Clone opensaas. We are using opensaas.sh directly.
2. Change the code as [doc](https://docs.opensaas.sh/guides/authentication/#email-verified-auth) said
3. Change user entity on main.wasp, like  
```
entity User {=psl
  id                        Int             @id @default(autoincrement())
  email                     String?         @unique
  username                  String          @unique
  password                  String?
  createdAt                 DateTime        @default(now())
...
```
5. Sign up on the browser
6. See error on the console: 
```
[Server]  src/auth/providers/email/signup.ts(48,39): error TS2345: Argument of type '{ email: any; password: any; }' is not assignable to parameter of type 'UserCreateInput'.
[Server]    Property 'username' is missing in type '{ email: any; password: any; }' but required in type 'UserCreateInput'.
```

And If I change the code wasp generated in `.wasp/out/server/src/auth/providers/email/signup.ts` from 
```
        const user = await createUser({
            ...additionalFields,
            email: userFields.email,
            password: userFields.password,
        });

```
to 

```
        const user = await createUser({
            ...additionalFields,
            email: userFields.email,
            username: userFields.email,
            password: userFields.password,
        });
```

The user signed up successfully.

**Expected behavior**
When I change the auth to email auth, the user can sign up correctly.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: MacOS 11.7.4
 - Browser Chrome
 - Version 120.0.6099.129

**Additional context**

**Comment by Martinsos** [comment]
I moved this issue to open-saas repo, as that makes more sense for it!

So the problem here @zacker330 is that `username` field is not optional on your Prisma model. If you make it optional (with `?`), then this will work.
However, Wasp by default will not set it to email, instead it will leave the username empty.

**Comment by zacker330** [comment]
> I moved this issue to open-saas repo, as that makes more sense for it!
> 
> So the problem here @zacker330 is that `username` field is not optional on your Prisma model. If you make it optional (with `?`), then this will work. However, Wasp by default will not set it to email, instead it will leave the username empty.

Thanks for Martinsos. 

Let me first explain the complete process.

## The Complete Process
Originally, I did set username as a optional field, email and password are required as the following code:
```
  email                     String         @unique
  username                  String?          @unique
  password                  String
```
But I got the compile error after I run `wasp build` or `wasp db migrate-dev`, like this:
```
❌ --- [Error] Your wasp project failed to compile: -------------------------------

- Entity 'User' (referenced by app.auth.userEntity) must have field 'email' of type 'String'.
- Entity 'User' (referenced by app.auth.userEntity) must have field 'password' of type 'String'.
```
Here is the example repo for this issue: https://github.com/zacker330/opensaas-test/commit/7b97dc2873115e80a26feae3bb7296b5a7a69074

When I changed the field to the oppsite, the build is pass. Here the commit for this change https://github.com/zacker330/opensaas-test/commit/66815374a98cf33e5e80be2e3d4f8dccbecc91cd: 
```
  email                     String?         @unique
  username                  String          @unique
  password                  String?
```

But when I run `wasp start`, I got another error:

```
[Server]  > server@0.0.0 build
[Server]  > npx tsc
[Server]
[Server]  src/auth/providers/email/signup.ts(48,39): error TS2345: Argument of type '{ email: any; password: any; }' is not assignable to parameter of type 'UserCreateInput'.
[Server]    Property 'username' is missing in type '{ email: any; password: any; }' but required in type 'UserCreateInput'.
[Server]  [nodemon] app crashed - waiting for file changes before starting...
```
 So, I attempted to fix this new issue by changing .wasp/out/server/src/auth/providers/email/signup.ts file directly as I'd said  previously. 

I noticed that @Martinsos  said  leaving the username empty. But it would not be working.  username is used by so many places, like UI and UserCreateInput type. 

## The Root Cause
I think the root cause is that:
1. The compile error prove that wasp read User entity incorrectly.
2. The email auth logic that wasp generated is not covering the use case of its.

**Comment by Martinsos** [comment]
Hm this is very weird @zacker330 , and thanks for all the info!

Regarding root causes that you listed:
1. I don't think it is likely/possible that Wasp wrongly read the User entity -> something else must be happening there.
2. It is not that logic that wasp generated is not covering the use case, it is that you can't have a required username in this case, because it doesn't make sense for how currently auth works in Wasp, which is what error says (although in a bit convoluted way).

What I would like to understand is what is causing that first error that you get, with missing pass/email, we will investigate this. This example you provided should help for sure.

As for leaving the username empty -> aha, it is used in many places hm! But what you can do is just use email instead in those places, if you don't have the username, right? Anyway you wanted to set the username to email, so it is the same thing.

@vincanger what are your thoughts on this?

**Comment by vincanger** [comment]
Hey @zacker330,

I believe that there actually is a bug in the Wasp Auth in this case. @infomiho can confirm if this is true or not.

For the time being, you can get your code to compile and run by setting up your User Entity like so:

```js
entity User {=psl
  id                   Int                @id @default(autoincrement())
  email                String?        @unique
  username             String?         @unique
  password             String?
//...
```

let me know if that works!

and as @Martinsos mentioned, if the client code is looking for a `username` property but it's undefined, then just change it to use the `email` property.

or if you want to programmatically set a `username` for each user on sign up, then you can customize the signup code like so:

```js 
// main.wasp

app crudTesting {
  // ...
  auth: {
    userEntity: User,
    methods: {
      email: {}
    },
    onAuthFailedRedirectTo: "/login",
    signup: {
      additionalFields: import { fields } from "@server/auth/signup.js",
    },
  },
}
```

```js
// signup.js

import { defineAdditionalSignupFields } from '@wasp/auth/index.js';

export default defineAdditionalSignupFields({
  username: (data) => {
    return data.email.split('@')[0]
  },
});
```

the code above will set first part of the name from the user's email as their `username`, e.g. if user signs up with 'zacker@gmail.com', the username would be `zacker`.

**Comment by pearcy** [comment]
Not sure if this is related, just in case - when I spin up a default repo (https://github.com/wasp-lang/open-saas.git) everything works swimmingly and everything works as expected except getting access to the Dashboard as an admin user. 

env changes:
// app/.env.server
'ADMIN_EMAILS=my@email.com'

// app/main.wasp
 ' isAdmin                   Boolean         @default(true)'

Expected behavior:
After these env changes, I signup and should have access to the Dashboard. 
-- there is nothing in my account in nav so enter /admin directly in the URL. 

If I change all the values to "false", I can get access to the Dashboard but when I go back to account or demo-app it routes me to the login screen. 
// app/main.wasp

`route AdminRoute { path: "/admin", to: DashboardPage }
page DashboardPage {
  authRequired: true,
  component: import Dashboard from "@client/admin/pages/DashboardPage"
}`

as a test I tried a to change a few settings but still experienced the same effect:
// app/src/client/admin/pages/DashboardPage.tsx
`
const Dashboard = ({ user }: { user: User }) => {
  const history = useHistory();
  if (user && !user.isAdmin) {
    history.push('/');
  }'

**Comment by vincanger** [comment]
> Same problem here

@MiqueiasGFernandes try making the `email`, `username`, and `password` fields of your user entity optional (`?`), like this:

```
entity User {=psl
  id                        Int             @id @default(autoincrement())
  email                     String?         @unique
  username                  String?         @unique
  password                  String?
  //...
  ```

**Comment by danjeldurmo** [comment]
Hi all,

I've had the same problem. If i put:

`entity User {=psl
  id                        Int             @id @default(autoincrement())
  email                     String?         @unique
  username                  String?         @unique
  password                  String?
  //...` 

App ask me for a username

<img width="467" alt="Screenshot 2024-02-06 alle 17 43 56" src="https://github.com/wasp-lang/open-saas/assets/66755329/9b03b0c6-6f0d-4dc9-84c2-48b73125f630">

where i can put this

**Comment by danjeldurmo** [comment]
If I remove "?" from username i i'm facing with this problem 
<img width="457" alt="Screenshot 2024-02-06 alle 18 03 26" src="https://github.com/wasp-lang/open-saas/assets/66755329/0078d7c4-9dec-4894-bd1f-dcc00a6b5885">

**Comment by MoeedthePessimist** [comment]
I am also getting the same error.

![image](https://github.com/wasp-lang/open-saas/assets/63580699/d237b716-0240-4e85-8df6-7ead195aa283)

I have already gone through the documentation and followed all the steps

![image](https://github.com/wasp-lang/open-saas/assets/63580699/72eb9718-c563-4cdf-a2a2-1716f7694fcb)

![image](https://github.com/wasp-lang/open-saas/assets/63580699/7b793fd9-2623-4475-8641-8a0adb733333)

Is there any quick patch for this issue?

**Comment by vincanger** [comment]
hey @MiqueiasGFernandes & @zacker330 

if you pull the newest version of the template which runs on Wasp v0.12 has updated Authentication methods! So migrating to new auth methods shouldn't be an issue.

give it a try here https://docs.opensaas.sh/start/getting-started/#cloning-the-opensaas-template

I'm closing this issue for now, but feel free to respond here if you have any questions and I'll reopen it :)
